File: mobile/src/main/java/com/nomi/artwatch/model/PostModel.java
package com.nomi.artwatch.model;

import com.tumblr.jumblr.types.Blog;
import com.tumblr.jumblr.types.Photo;
import com.tumblr.jumblr.types.PhotoPost;
import com.tumblr.jumblr.types.Post;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.schedulers.Schedulers;
import rx.subjects.AsyncSubject;
import timber.log.Timber;

/**
 * Created by Ryota Niinomi on 2015/11/08.
 */
@Singleton
public class PostModel extends BaseModel {

    private static final int LIMIT = 40;

    @Inject
    public PostModel() {

    }

    public Observable<Blog> getBlog(String blogName) {
        AsyncSubject<Blog> subject = AsyncSubject.create();

        Observable
                .just(null)
                .subscribeOn(Schedulers.io())
                .subscribe(aVoid -> {
                    Blog blog = getClient().blogInfo(blogName);

                    subject.onNext(blog);

                }, throwable -> {
                    Timber.w(throwable, throwable.getLocalizedMessage());
                    subject.onError(throwable);

                }, subject::onCompleted);

        return subject;
    }

    public Observable<Void> getPost(String blogName) {
        AsyncSubject<Void> subject = AsyncSubject.create();

        Observable
                .just(null)
                .subscribeOn(Schedulers.io())
                .subscribe(aVoid -> {
                    List<Post> blogPosts = getClient().blogPosts(blogName);

                    subject.onNext(null);

                }, throwable -> {
                    Timber.w(throwable, throwable.getLocalizedMessage());
                    subject.onError(throwable);

                }, subject::onCompleted);

        return subject;
    }

    public Observable<List<Photo>> getPhotoPost(String blogName) {
        AsyncSubject<List<Photo>> subject = AsyncSubject.create();

        Observable
                .just(null)
                .subscribeOn(Schedulers.io())
                .subscribe(aVoid -> {
                    List<PhotoPost> photoPosts = getClient ().photos(blogName, null);
                    List<Photo> photos = toPhotos(photoPosts);

                    subject.onNext(photos);

                }, throwable -> {
                    Timber.w(throwable, throwable.getLocalizedMessage());
                    subject.onError(throwable);

                }, subject::onCompleted);

        return subject;
    }

    private List<Photo> toPhotos(List<PhotoPost> photoPosts) {
        List<Photo> photos = new ArrayList<>();

        Observable
                .from(photoPosts)
                .forEach(item -> {
                    Observable
                            .from(item.getPhotos())
                            .filter(item2 -> photos.size() < LIMIT && isGif(item2))
                            .forEach(item3 -> {
                                Timber.d("photo url : " + item3.getOriginalSize().getUrl());
                                photos.add(item3);
                            });
                });

        return photos;
    }

    private boolean isGif(Photo photo) {
        return photo.getOriginalSize().getUrl().contains(".gif");
    }
}

File: mobile/src/main/kotlin/com/nomi/artwatch/model/PostModel.kt
package com.nomi.artwatch.model

import android.content.Context
import com.nomi.artwatch.R
import com.tumblr.jumblr.types.Blog
import com.tumblr.jumblr.types.Photo
import com.tumblr.jumblr.types.PhotoPost
import rx.Observable
import rx.schedulers.Schedulers
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Ryota Niinomi on 2015/11/08.
 */
@Singleton
class PostModel
@Inject
constructor(private val mContext: Context, prefModel: PrefModel) : BaseModel(prefModel) {

    companion object {
        private val LIMIT = 40
    }

    fun getBlog(blogName: String): Observable<Blog> = Observable
            .just(null)
            .subscribeOn(Schedulers.io())
            .map { return@map client.blogInfo(blogName) }

    fun getPost(blogName: String): Observable<Void> = Observable
            .just(null)
            .subscribeOn(Schedulers.io())
            .map{
                client.blogPosts(blogName)
                return@map null
            }

    fun getPhotoPost(blogName: String): Observable<List<Photo>> = Observable
            .just(null)
            .subscribeOn(Schedulers.io())
            .map {
                val photoPosts = client.photos(blogName, null)
                val photos = toPhotos(photoPosts)
                return@map photos
            }
            .onErrorReturn {
                Timber.e(it, it.message)
                throw RuntimeException(mContext.getString(R.string.error_common))
            }

    private fun toPhotos(photoPosts: List<PhotoPost>): List<Photo> {
        val photos = ArrayList<Photo>()

        Observable
                .from(photoPosts)
                .forEach { item ->
                    Observable
                            .from(item.photos)
                            .filter { item2 -> photos.size < LIMIT && isGif(item2) }
                            .forEach { item3 ->
                                Timber.d("photo url : " + item3.originalSize.url)
                                photos.add(item3)
                            }
                }

        return photos
    }

    private fun isGif(photo: Photo): Boolean {
        return photo.originalSize.url.contains(".gif")
    }
}
